lib/bloc/nostr/nostr_bloc.dart

```diff
@@ -0,0 +1,168 @@
+import 'package:nostr_tools/nostr_tools.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+class NostrBloc with AsyncActionsHandler {
+  String nostrPublicKey;
+  String nostrPrivateKey;
+
+  FlutterSecureStorage _secureStorage;
+  SharedPreferences sharedPreferences;
+
+  NostrBloc() {
+    _secureStorage = const FlutterSecureStorage();
+
+    initNostr();
+
+    registerAsyncHandlers({
+      GetPublicKey: _handleGetPublicKey,
+      SignEvent: _handleSignEvent,
+      GetRelays: _handleGetRelays,
+      Nip04Encrypt: _handleNip04Encrypt,
+      Nip04Decrypt: _handleNip04Decrypt,
+    });
+    listenActions();
+  }
+
+  void initNostr() async {
+    nostrPublicKey = await _secureStorage.read(key: 'nostrPublicKey');
+    nostrPrivateKey = await _secureStorage.read(key: 'nostrPrivateKey');
+    SharedPreferences sharedPreferences = await SharedPreferences.getInstance();
+
+    if (nostrPublicKey == null) {
+      sharedPreferences.setBool('rememberGetPubKeyChoice', false);
+      sharedPreferences.setBool('rememberSignEventChoice', false);
+    }
+  }
+
+  final StreamController<String> _publicKeyController =
+      StreamController<String>.broadcast();
+  Stream<String> get publicKeyStream => _publicKeyController.stream;
+
+  final StreamController<Map<String, dynamic>> _eventController =
+      StreamController<Map<String, dynamic>>.broadcast();
+
+  Stream<Map<String, dynamic>> get eventStream => _eventController.stream;
+
+  Future<void> _handleGetPublicKey(GetPublicKey action) async {
+    //  to get the public key
+    String publicKey = await _fetchPublicKey();
+    _publicKeyController.add(publicKey);
+    action.resolve(publicKey);
+  }
+
+  Future<void> _handleSignEvent(SignEvent action) async {
+    //  to sign the event
+    Map<String, dynamic> signedEvent =
+        await _signEvent(action.eventObject, action.privateKey);
+
+    _eventController.add(signedEvent);
+    action.resolve(signedEvent);
+  }
+
+  Future<void> _handleGetRelays(GetRelays action) async {
+    //  to get the relays
+    List<String> relays = await _fetchRelays();
+    // Perform any processing on the relays if needed
+    action.resolve(relays);
+  }
+
+  Future<void> _handleNip04Encrypt(Nip04Encrypt action) async {
+    // to encrypt the data`
+    String encryptedData = await _encryptData(action.data, action.publicKey);
+    action.resolve(encryptedData);
+  }
+
+  Future<void> _handleNip04Decrypt(Nip04Decrypt action) async {
+    // to decrypt the data
+    String decryptedData =
+        await _decryptData(action.encryptedData, action.privateKey);
+    action.resolve(decryptedData);
+  }
+
+  // Methods to simulate the actual logic
+
+  Future<String> _fetchPublicKey() async {
+    // check if key pair already exists otherwise generate it
+
+    if (nostrPublicKey == null) {
+      final keyGenerator = KeyApi();
+      nostrPrivateKey = keyGenerator.generatePrivateKey();
```

Comment:
This key generator is not reproducible if the user restores her wallet.
If we want to make it reproducible we need to save it in the breez.db (on the golang side).

Reply:
Okay, so that means **key pairs** once generated should be saved in breez.db right?